pennylane_lightning/core/src/measurements/MeasurementsBase.hpp

```diff
@@ -111,6 +114,183 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
+                 obs_term_idx++) {
+                auto obs_samples = measure_with_samples(
+                    obs, num_shots, shot_range, obs_term_idx);
+                PrecisionT result_per_term = std::accumulate(
+                    obs_samples.begin(), obs_samples.end(), 0.0);
+
+                result +=
+                    coeffs[obs_term_idx] * result_per_term / obs_samples.size();
+            }
+        } else {
+            auto obs_samples = measure_with_samples(obs, num_shots, shot_range);
+            result =
+                std::accumulate(obs_samples.begin(), obs_samples.end(), 0.0);
+            result = result / obs_samples.size();
+        }
+        return result;
+    }
+
+    /**
+     * @brief Calculate the expectation value for a general Observable.
+     *
+     * @param obs Observable.
+     * @param num_shots Number of shots used to generate samples
+     * @param shot_range The range of samples to use. All samples are used
+     * by default.
+     * @param term_idx Index of a Hamiltonian term
+     *
+     * @return Expectation value with respect to the given observable.
+     */
+    auto measure_with_samples(const Observable<StateVectorT> &obs,
+                              const size_t &num_shots,
+                              const std::vector<size_t> &shot_range,
+                              size_t term_idx = 0) {
+        const size_t num_qubits = _statevector.getTotalNumQubits();
+        std::vector<size_t> obs_wires;
+        std::vector<size_t> identity_wires;
+
+        auto sub_samples = _sample_state(obs, num_shots, shot_range, obs_wires,
+                                         identity_wires, term_idx);
+
+        size_t num_samples = num_shots;
+        if (!shot_range.empty()) {
+            num_samples = shot_range.size();
+        }
+        std::vector<PrecisionT> obs_samples(num_samples, 0);
+
+        size_t num_identity_obs = identity_wires.size();
+        if (!identity_wires.empty()) {
+            size_t identity_obs_idx = 0;
+            for (size_t i = 0; i < obs_wires.size(); i++) {
+                if (identity_wires[identity_obs_idx] == obs_wires[i]) {
+                    std::swap(obs_wires[identity_obs_idx], obs_wires[i]);
+                    identity_obs_idx++;
+                }
+            }
+        }
+
+        for (size_t i = 0; i < num_samples; i++) {
+            std::vector<size_t> local_sample(obs_wires.size());
+            size_t idx = 0;
+            for (auto &obs_wire : obs_wires) {
+                local_sample[idx] = sub_samples[i * num_qubits + obs_wire];
+                idx++;
+            }
+
+            if (num_identity_obs != obs_wires.size()) {
+                // eigen values are `1` and `-1` for PauliX, PauliY, PauliZ,
+                // Hadamard gates the eigen value for a eigen vector |00001> is
+                // -1 since sum of the value at each bit position is odd
+                if ((static_cast<size_t>(std::accumulate(
```

Comment:
Could you extract out the accumulate from the if-condition? This would simplify the condition as well as the code style.

Reply:
Sure!